File: tai/commands/docs.py
"""tai docs — documentation utilities for skills."""

from __future__ import annotations

import json
import re
from pathlib import Path

DOCS_DIR_NAME = "docs"
ASSETS_DIR = "_assets"

# Bundled assets shipped with tai
_BUNDLED_ASSETS = Path(__file__).resolve().parent.parent / "data" / "docs" / "assets"


# ── Validation schemas ──────────────────────────────────────────────────────

SCHEMAS: dict[str, dict] = {
    "intent": {
        "required_sections": ["context", "problem", "solution", "success-criteria"],
        "required_meta": ["date"],
    },
    "decision": {
        "required_sections": ["context", "decision", "consequences"],
        "required_meta": ["date", "status"],
    },
    "design": {
        "required_sections": ["overview", "components"],
        "required_meta": ["date"],
    },
    "spec": {
        "required_sections": ["problem", "requirements", "acceptance-criteria"],
        "required_meta": ["date", "status"],
    },
    "guide": {
        "required_sections": ["overview"],
        "required_meta": ["date"],
    },
    "plan": {
        "required_sections": ["phases"],
        "required_meta": ["date"],
    },
    "review": {
        "required_sections": ["findings"],
        "required_meta": ["date"],
    },
    "trace": {
        "required_sections": [],
        "required_meta": ["date"],
    },
    "changelog": {
        "required_sections": [],
        "required_meta": ["date"],
    },
}


def find_docs_root() -> Path:
    """Walk up from cwd to find the docs/ directory."""
    cwd = Path.cwd()
    for parent in [cwd, *cwd.parents]:
        candidate = parent / DOCS_DIR_NAME
        if candidate.is_dir():
            return candidate
        if (parent / ".git").exists():
            return candidate
    return cwd / DOCS_DIR_NAME


def discover_docs(docs_root: Path) -> list[dict]:
    """Find all .html docs under docs_root, extract title and metadata."""
    docs = []
    if not docs_root.is_dir():
        return docs

    for html_file in sorted(docs_root.rglob("*.html")):
        rel = html_file.relative_to(docs_root)
        if rel.parts[0] == ASSETS_DIR:
            continue

        text = html_file.read_text(encoding="utf-8", errors="replace")
        title = _extract_title(text) or rel.stem
        doc_type = _extract_meta(text, "doc-type") or ""

        docs.append({
            "path": str(rel),
            "title": title,
            "type": doc_type,
        })
    return docs


def validate_file(path: Path, docs_root: Path) -> list[str]:
    """Validate a single HTML doc. Returns list of issues."""
    text = path.read_text(encoding="utf-8", errors="replace")
    issues: list[str] = []

    title = _extract_title(text)
    doc_type = _extract_meta(text, "doc-type")
    doc_date = _extract_meta(text, "doc-date")

    if not title or title == "Untitled":
        issues.append("Missing <title>")
    if not doc_type:
        issues.append('Missing <meta name="doc-type">')
    if not doc_date:
        issues.append('Missing <meta name="doc-date">')

    schema = SCHEMAS.get(doc_type or "")
    if schema:
        for section in schema["required_sections"]:
            if f'data-section="{section}"' not in text:
                issues.append(f"Missing section: {section}")
        for meta_field in schema["required_meta"]:
            if not _extract_meta(text, f"doc-{meta_field}"):
                issues.append(f"Missing meta: doc-{meta_field}")
    elif doc_type:
        issues.append(f"Unknown doc-type: {doc_type}")

    for href in _extract_internal_links(text):
        target = (path.parent / href).resolve()
        if not target.exists():
            issues.append(f"Broken link: {href}")

    return issues


def validate_all(docs_root: Path) -> dict[str, list[str]]:
    """Validate every HTML doc under docs_root. Returns {path: [issues]}."""
    results: dict[str, list[str]] = {}
    if not docs_root.is_dir():
        return results

    for html_file in sorted(docs_root.rglob("*.html")):
        rel = html_file.relative_to(docs_root)
        if rel.parts[0] == ASSETS_DIR:
            continue
        file_issues = validate_file(html_file, docs_root)
        if file_issues:
            results[str(rel)] = file_issues
    return results


# ── HTML parsing helpers (no external deps) ─────────────────────────────────

_RE_TITLE = re.compile(r"<title[^>]*>([^<]+)</title>", re.IGNORECASE)
_RE_META = re.compile(r'<meta\s+name="([^"]+)"\s+content="([^"]*)"', re.IGNORECASE)
_RE_HREF = re.compile(r'<a\s[^>]*href="([^"#][^"]*)"', re.IGNORECASE)


def _extract_title(html: str) -> str | None:
    m = _RE_TITLE.search(html)
    return m.group(1).strip() if m else None


def _extract_meta(html: str, name: str) -> str | None:
    for m in _RE_META.finditer(html):
        if m.group(1) == name:
            return m.group(2)
    return None


def _extract_internal_links(html: str) -> list[str]:
    links = []
    for m in _RE_HREF.finditer(html):
        href = m.group(1)
        if href.startswith(("http://", "https://", "mailto:", "/")):
            continue
        links.append(href)
    return links


# ── Asset management ────────────────────────────────────────────────────────

def _copy_assets(target: Path) -> None:
    """Copy bundled CSS/JS assets to the project's docs/_assets/."""
    import shutil
    if target.is_dir():
        shutil.rmtree(target)
    shutil.copytree(_BUNDLED_ASSETS, target)


def write_index(docs_root: Path) -> None:
    """Inject a static <nav> sidebar into every HTML file. No JS needed."""
    docs = discover_docs(docs_root)

    # Group by directory
    groups: dict[str, list[dict]] = {}
    for doc in docs:
        parts = doc["path"].split("/")
        d = parts[0] if len(parts) > 1 else ""
        groups.setdefault(d, []).append(doc)

    for html_file in sorted(docs_root.rglob("*.html")):
        rel = html_file.relative_to(docs_root)
        if rel.parts[0] == ASSETS_DIR:
            continue

        current = str(rel)
        depth = len(rel.parts) - 1
        prefix = "../" * depth

        # Build nav HTML
        lines = ['<nav class="docs-nav">']
        lines.append(f'  <a class="nav-title" href="{prefix}index.html">Docs</a>')

        for d in sorted(groups, key=lambda x: (x != "", x)):
            if d:
                lines.append(f'  <div class="nav-group">{d}</div>')
                lines.append('  <div class="nav-group-items">')

            for doc in sorted(groups[d], key=lambda x: x["title"]):
                href = prefix + doc["path"]
                active = " active" if doc["path"] == current else ""
                lines.append(
                    f'    <a class="nav-link{active}" href="{href}">{doc["title"]}</a>'
                )

            if d:
                lines.append('  </div>')

        lines.append('</nav>')
        nav_html = "\n".join(lines)

        text = html_file.read_text(encoding="utf-8")

        # Remove old nav if present
        text = re.sub(
            r'<nav class="docs-nav">.*?</nav>\s*',
            '',
            text,
            flags=re.DOTALL,
        )

        # Normalize body tag and insert nav
        text = re.sub(r'<body[^>]*>', '<body class="has-nav">', text, count=1)
        text = text.replace('<body class="has-nav">', f'<body class="has-nav">\n{nav_html}', 1)

        html_file.write_text(text, encoding="utf-8")

File: docs/_assets/docs.js
/* ── tai docs — runtime ──────────────────────────────────────────────────
   Simple: fetch file tree from server, build sidebar, search, navigate.
   All links absolute. No broken paths.
   ────────────────────────────────────────────────────────────────────── */

(function () {
  "use strict";

  var served = location.protocol.startsWith("http");

  /* ── Sidebar ──────────────────────────────────────────────────────── */

  function buildSidebar() {
    /* try server API first, fall back to static index file */
    var urls = served
      ? ["/_api/docs", "_assets/_index.json"]
      : ["_assets/_index.json", "../_assets/_index.json"];

    fetchFirst(urls).then(function (docs) {
      if (!docs) return;
        var nav = document.createElement("nav");
        nav.className = "docs-nav";

        /* title */
        var title = document.createElement("a");
        title.className = "nav-title";
        title.href = "/";
        title.textContent = "Docs";
        nav.appendChild(title);

        /* search */
        var search = document.createElement("input");
        search.className = "nav-search";
        search.type = "text";
        search.placeholder = "Search...";
        search.addEventListener("input", function () {
          var q = search.value.toLowerCase();
          nav.querySelectorAll(".nav-link").forEach(function (a) {
            a.style.display = a.textContent.toLowerCase().includes(q) ? "" : "none";
          });
          nav.querySelectorAll(".nav-group").forEach(function (g) {
            var links = g.nextElementSibling;
            if (!links || !links.classList.contains("nav-group-items")) return;
            var anyVisible = false;
            links.querySelectorAll(".nav-link").forEach(function (a) {
              if (a.style.display !== "none") anyVisible = true;
            });
            g.style.display = anyVisible || !q ? "" : "none";
            links.style.display = anyVisible || !q ? "" : "none";
          });
        });
        nav.appendChild(search);

        /* back button — show parent dir */
        var current = location.pathname.replace(/^\//, "");
        var parts = current.split("/");
        if (parts.length > 1) {
          var back = document.createElement("a");
          back.className = "nav-back";
          back.href = "/" + parts.slice(0, -1).join("/") + "/";
          back.textContent = "\u2190 " + (parts.length > 2 ? parts[parts.length - 2] : "root");
          nav.appendChild(back);
        }

        /* build tree: group by first directory */
        var tree = {};
        docs.forEach(function (doc) {
          var segs = doc.path.split("/");
          var dir = segs.length > 1 ? segs[0] : "";
          if (!tree[dir]) tree[dir] = [];
          tree[dir].push(doc);
        });

        /* render root files first, then directories */
        var dirs = Object.keys(tree).sort(function (a, b) {
          if (a === "") return -1;
          if (b === "") return 1;
          return a.localeCompare(b);
        });

        dirs.forEach(function (dir) {
          if (dir) {
            var heading = document.createElement("div");
            heading.className = "nav-group";
            heading.textContent = dir;
            heading.addEventListener("click", function () {
              var items = heading.nextElementSibling;
              if (items) items.classList.toggle("collapsed");
              heading.classList.toggle("collapsed");
            });
            nav.appendChild(heading);

            var container = document.createElement("div");
            container.className = "nav-group-items";
            /* auto-expand current directory */
            if (!current.startsWith(dir + "/")) {
              container.classList.add("collapsed");
              heading.classList.add("collapsed");
            }

            tree[dir]
              .sort(function (a, b) { return a.title.localeCompare(b.title); })
              .forEach(function (doc) { container.appendChild(makeLink(doc, current)); });
            nav.appendChild(container);
          } else {
            tree[dir]
              .sort(function (a, b) { return a.title.localeCompare(b.title); })
              .forEach(function (doc) { nav.appendChild(makeLink(doc, current)); });
          }
        });

        document.body.prepend(nav);
        document.body.classList.add("has-nav");
      })
    });
  }

  function fetchFirst(urls) {
    var i = 0;
    function tryNext() {
      if (i >= urls.length) return Promise.resolve(null);
      return fetch(urls[i++])
        .then(function (r) { return r.ok ? r.json() : tryNext(); })
        .catch(function () { return tryNext(); });
    }
    return tryNext();
  }

  function makeLink(doc, current) {
    var a = document.createElement("a");
    a.href = "/" + doc.path;
    a.className = "nav-link";
    a.textContent = doc.title;
    if (current === doc.path) a.classList.add("active");
    if (doc.type) a.dataset.type = doc.type;
    return a;
  }

  /* ── Metadata display ─────────────────────────────────────────────── */

  function renderMeta() {
    var article = document.querySelector("article");
    if (!article || article.querySelector("dl.meta")) return;

    var docType = getMeta("type");
    var statusTypes = ["spec", "decision"];
    var fields = ["type", "status", "date", "author"];
    var entries = [];

    fields.forEach(function (f) {
      var v = getMeta(f);
      if (!v) return;
      if (f === "status" && statusTypes.indexOf(docType) === -1) return;
      entries.push([f, v]);
    });

    if (!entries.length) return;

    var dl = document.createElement("dl");
    dl.className = "meta";
    entries.forEach(function (pair) {
      var dt = document.createElement("dt");
      dt.textContent = pair[0];
      var dd = document.createElement("dd");
      if (pair[0] === "status") {
        var span = document.createElement("span");
        span.className = "status status-" + pair[1];
        span.textContent = pair[1];
        dd.appendChild(span);
      } else {
        dd.textContent = pair[1];
      }
      dl.appendChild(dt);
      dl.appendChild(dd);
    });

    var h1 = article.querySelector("h1");
    if (h1 && h1.nextSibling) h1.parentNode.insertBefore(dl, h1.nextSibling);
    else article.prepend(dl);
  }

  /* ── Validation ───────────────────────────────────────────────────── */

  var SCHEMAS = {
    intent:    { req: ["context", "problem", "solution", "success-criteria"], meta: ["date"] },
    decision:  { req: ["context", "decision", "consequences"], meta: ["date", "status"] },
    design:    { req: ["overview", "components"], meta: ["date"] },
    spec:      { req: ["problem", "requirements", "acceptance-criteria"], meta: ["date", "status"] },
    guide:     { req: ["overview"], meta: ["date"] },
    plan:      { req: ["phases"], meta: ["date"] },
    review:    { req: ["findings"], meta: ["date"] },
    trace:     { req: [], meta: ["date"] },
    changelog: { req: [], meta: ["date"] },
  };

  function validate() {
    var article = document.querySelector("article");
    if (!article) return;

    var issues = [];
    var docType = getMeta("type");

    if (!document.title || document.title === "Untitled") issues.push("Missing <title>");
    if (!docType) issues.push("Missing doc-type meta");
    if (!getMeta("date")) issues.push("Missing doc-date meta");

    var schema = SCHEMAS[docType];
    if (schema) {
      schema.req.forEach(function (s) {
        if (!article.querySelector('[data-section="' + s + '"]'))
          issues.push("Missing section: " + s);
      });
      schema.meta.forEach(function (m) {
        if (!getMeta(m)) issues.push("Missing meta: doc-" + m);
      });
    }

    /* badge */
    var badge = document.createElement("div");
    badge.className = "validation-badge " + (issues.length ? "validation-fail" : "validation-pass");
    badge.textContent = issues.length ? issues.length + " issue" + (issues.length > 1 ? "s" : "") : "\u2713 Valid";
    badge.title = issues.join("\n") || "All checks passed";

    var panel = document.createElement("div");
    panel.className = "validation-panel";
    issues.forEach(function (msg) {
      var div = document.createElement("div");
      div.className = "issue";
      div.textContent = msg;
      panel.appendChild(div);
    });
    if (!issues.length) {
      var ok = document.createElement("div");
      ok.style.color = "var(--success)";
      ok.textContent = "All checks passed.";
      panel.appendChild(ok);
    }

    badge.addEventListener("click", function () { panel.classList.toggle("open"); });
    document.addEventListener("click", function (e) {
      if (!badge.contains(e.target) && !panel.contains(e.target)) panel.classList.remove("open");
    });

    document.body.appendChild(panel);
    document.body.appendChild(badge);
  }

  /* ── Hot reload ───────────────────────────────────────────────────── */

  function hotReload() {
    if (!served) return;
    var es = new EventSource("/_events");
    es.onmessage = function (e) {
      try { if (JSON.parse(e.data).type === "reload") location.reload(); } catch (_) {}
    };
    es.onerror = function () { es.close(); setTimeout(hotReload, 2000); };
  }

  /* ── Theme ────────────────────────────────────────────────────────── */

  function initTheme() {
    if (localStorage.getItem("tai-docs-theme") === "dark")
      document.documentElement.setAttribute("data-theme", "dark");

    var btn = document.createElement("button");
    btn.className = "theme-toggle";
    btn.setAttribute("aria-label", "Toggle dark mode");
    btn.textContent = isDark() ? "\u2600" : "\u263E";
    btn.addEventListener("click", function () {
      var dark = !isDark();
      document.documentElement.setAttribute("data-theme", dark ? "dark" : "light");
      localStorage.setItem("tai-docs-theme", dark ? "dark" : "light");
      btn.textContent = dark ? "\u2600" : "\u263E";
    });
    document.body.appendChild(btn);
  }

  function isDark() { return document.documentElement.getAttribute("data-theme") === "dark"; }

  /* ── Fonts ────────────────────────────────────────────────────────── */

  function loadFonts() {
    if (document.querySelector("link[data-tai-fonts]")) return;
    var link = document.createElement("link");
    link.rel = "stylesheet";
    link.dataset.taiFonts = "true";
    link.href = "https://fonts.googleapis.com/css2?family=VT323&family=Source+Serif+4:ital,opsz,wght@0,8..60,400;0,8..60,700;1,8..60,400&family=JetBrains+Mono:wght@400;500;600&display=swap";
    document.head.appendChild(link);
  }

  /* ── Helpers ──────────────────────────────────────────────────────── */

  function getMeta(name) {
    var el = document.querySelector('meta[name="doc-' + name + '"]');
    return el ? el.content : null;
  }

  /* ── Init ──────────────────────────────────────────────────────────── */

  document.addEventListener("DOMContentLoaded", function () {
    loadFonts();
    renderMeta();
    buildSidebar();
    validate();
    hotReload();
    initTheme();
  });
})();
